File: vb.net_samples/CreateTicketWithAttachment.vb
Imports System.IO
Imports System.Net
Imports System.Text

Namespace FreshdeskTest
	Class CreateTicketWithAttachment
		Private Const _APIKey As String = "dmOra0rbCOgI3R9lPna"
		Private Const _Url As String = "http://googlehelper.freshdesk.com/helpdesk/tickets.json"
		' verify if you have to use http or https for your account
		Private Shared Sub writeCRLF(o As Stream)
			Dim crLf As Byte() = Encoding.ASCII.GetBytes(vbCr & vbLf)
			o.Write(crLf, 0, crLf.Length)
		End Sub

		Private Shared Sub writeBoundaryBytes(o As Stream, b As String, isFinalBoundary As Boolean)
			Dim boundary As String = If(isFinalBoundary = True, "--" & b & "--", "--" & b & vbCr & vbLf)
			Dim d As Byte() = Encoding.ASCII.GetBytes(boundary)
			o.Write(d, 0, d.Length)
		End Sub

		Private Shared Sub writeContentDispositionFormDataHeader(o As Stream, name As String)
			Dim data As String = "Content-Disposition: form-data; name=""" & name & """" & vbCr & vbLf & vbCr & vbLf
			Dim b As Byte() = Encoding.ASCII.GetBytes(data)
			o.Write(b, 0, b.Length)
		End Sub

		Private Shared Sub writeContentDispositionFileHeader(o As Stream, name As String, fileName As String, contentType As String)
			Dim data As String = "Content-Disposition: form-data; name=""" & name & """; filename=""" & fileName & """" & vbCr & vbLf
			data += "Content-Type: " & contentType & vbCr & vbLf & vbCr & vbLf
			Dim b As Byte() = Encoding.ASCII.GetBytes(data)
			o.Write(b, 0, b.Length)
		End Sub

		Private Shared Sub writeString(o As Stream, data As String)
			Dim b As Byte() = Encoding.ASCII.GetBytes(data)
			o.Write(b, 0, b.Length)
		End Sub

		Public Shared Sub Main(args As String())
			Console.WriteLine("Application starting...")

			' Define boundary:
			Dim boundary As String = "----------------------------" & DateTime.Now.Ticks.ToString("x")

			' Web Request:
			Dim wr As HttpWebRequest = DirectCast(WebRequest.Create(_Url), HttpWebRequest)

			wr.Headers.Clear()

			' Method and headers:
			wr.ContentType = "multipart/form-data; boundary=" & boundary
			wr.Method = "POST"
			wr.KeepAlive = True

			' Basic auth:
			Dim login As String = _APIKey & ":X"
			Dim credentials As String = Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(login))
			wr.Headers(HttpRequestHeader.Authorization) = "Basic " & credentials

			' Body:
			Using rs As Stream = wr.GetRequestStream()
				' Email:
				writeBoundaryBytes(rs, boundary, False)
				writeContentDispositionFormDataHeader(rs, "helpdesk_ticket[email]")
				writeString(rs, "example@example.com")
				writeCRLF(rs)

				' Subject:
				writeBoundaryBytes(rs, boundary, False)
				writeContentDispositionFormDataHeader(rs, "helpdesk_ticket[subject]")
				writeString(rs, "Ticket Title")
				writeCRLF(rs)

				' Description:
				writeBoundaryBytes(rs, boundary, False)
				writeContentDispositionFormDataHeader(rs, "helpdesk_ticket[description]")
				writeString(rs, "Ticket description.")
				writeCRLF(rs)

				' Attachment:
				writeBoundaryBytes(rs, boundary, False)
				writeContentDispositionFileHeader(rs, "helpdesk_ticket[attachments][][resource]", "data.txt", "text/plain")
				Dim fs As New FileStream("/Users/user/Desktop/data.txt", FileMode.Open, FileAccess.Read)
				Dim data As Byte() = New Byte(fs.Length - 1) {}
				fs.Read(data, 0, data.Length)
				fs.Close()
				rs.Write(data, 0, data.Length)
				writeCRLF(rs)

				' End marker:
				writeBoundaryBytes(rs, boundary, True)

				rs.Close()
			End Using

			' Response processing:
			Try
				Console.WriteLine("Submitting Request")
				Dim response = DirectCast(wr.GetResponse(), HttpWebResponse)
				Dim resStream As Stream = response.GetResponseStream()
				Dim resJson As String = New StreamReader(resStream, Encoding.ASCII).ReadToEnd()
				Console.WriteLine(resJson)
			Catch ex As Exception
				Console.WriteLine("ERROR")
				Console.WriteLine(ex.Message)
			Finally
				Console.WriteLine(Environment.NewLine)
				Console.WriteLine(Environment.NewLine)
			End Try

		End Sub

	End Class
End Namespace

File: C-Sharp/CreateTicketWithAttachment.cs
using System;
using System.IO;
using System.Net;
using System.Text;

namespace FreshdeskTest
{
    class CreateTicketWithAttachment
    {

        private const string fdDomain = "YOUR_DOMAIN";
        private const string _APIKey = "YOUR_API_KEY";
        private const string path = "/api/v2/tickets";
        private const string _Url = "https://"+ fdDomain + ".freshdesk.com" + path; 
        
        private static void writeCRLF(Stream o)
        {
            byte[] crLf = Encoding.UTF8.GetBytes("\r\n");
            o.Write(crLf, 0, crLf.Length);
        }
        
        private static void writeBoundaryBytes(Stream o, string b, bool isFinalBoundary)
        {
            string boundary = isFinalBoundary == true ? "--" + b + "--" : "--" + b + "\r\n";
            byte[] d = Encoding.UTF8.GetBytes(boundary);
            o.Write(d, 0, d.Length);
        }
        
        private static void writeContentDispositionFormDataHeader(Stream o, string name)
        {
            string data = "Content-Disposition: form-data; name=\"" + name + "\"\r\n\r\n";
            byte[] b = Encoding.UTF8.GetBytes(data);
            o.Write(b, 0, b.Length);
        }
        
        private static void writeContentDispositionFileHeader(Stream o, string name, string fileName, string contentType)
        {
            string data = "Content-Disposition: form-data; name=\"" + name + "\"; filename=\"" + fileName + "\"\r\n";
            data += "Content-Type: " + contentType + "\r\n\r\n";
            byte[] b = Encoding.UTF8.GetBytes(data);
            o.Write(b, 0, b.Length);
        }
        
        private static void writeString(Stream o, string data)
        {
            byte[] b = Encoding.UTF8.GetBytes(data);
            o.Write(b, 0, b.Length);
        }

        static void Main(string[] args)
        {
            Console.WriteLine("Application starting...");

            // Define boundary:
            string boundary = "----------------------------" + DateTime.Now.Ticks.ToString("x");

            // Web Request:
            HttpWebRequest wr = (HttpWebRequest)WebRequest.Create(_Url);

            wr.Headers.Clear();
            
            // Method and headers:
            wr.ContentType = "multipart/form-data; boundary=" + boundary;
            wr.Method = "POST";
            wr.KeepAlive = true;
            
            // Basic auth:
            string login = _APIKey + ":X"; // It could be your username:password also.
            string credentials = Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(login));
            wr.Headers[HttpRequestHeader.Authorization] = "Basic " + credentials;
            
            // Body:
            using (var rs = wr.GetRequestStream())
            {
                // Email:
                writeBoundaryBytes(rs, boundary, false);
                writeContentDispositionFormDataHeader(rs, "email");
                writeString(rs, "example@example.com");
                writeCRLF(rs);
                
                // Subject:
                writeBoundaryBytes(rs, boundary, false);
                writeContentDispositionFormDataHeader(rs, "subject");
                writeString(rs, "Ticket Title");
                writeCRLF(rs);
                
                // Description:
                writeBoundaryBytes(rs, boundary, false);
                writeContentDispositionFormDataHeader(rs, "description");
                writeString(rs, "Ticket description.");
                writeCRLF(rs);

                // Status:
                writeBoundaryBytes(rs, boundary, false);
                writeContentDispositionFormDataHeader(rs, "status");
                writeString(rs, "2");
                writeCRLF(rs);

                // Priority:
                writeBoundaryBytes(rs, boundary, false);
                writeContentDispositionFormDataHeader(rs, "priority");
                writeString(rs, "2");
                writeCRLF(rs);
                
                // Attachment:
                writeBoundaryBytes(rs, boundary, false);
                writeContentDispositionFileHeader(rs, "attachments[]", "x.txt", "text/plain");
                FileStream fs = new FileStream("/path/to/my/file.txt", FileMode.Open, FileAccess.Read);
                byte[] data = new byte[fs.Length];
                fs.Read(data, 0, data.Length);
                fs.Close();
                rs.Write(data, 0, data.Length);
                writeCRLF(rs);
                
                // End marker:
                writeBoundaryBytes(rs, boundary, true);
                
                rs.Close();
            
                // Response processing:
                try
                {
                    Console.WriteLine("Submitting Request");
                    var response = (HttpWebResponse)wr.GetResponse();
                    Stream resStream = response.GetResponseStream();
                    string Response = new StreamReader(resStream, Encoding.UTF8).ReadToEnd();
                    //return status code
                    Console.WriteLine("Status Code: {1} {0}", ((HttpWebResponse)response).StatusCode, (int)((HttpWebResponse)response).StatusCode);
                    //return location header
                    Console.WriteLine("Location: {0}", response.Headers["Location"]);
                    //return the response 
                    Console.Out.WriteLine(Response);
                }
                catch (WebException ex)
                {
                    Console.WriteLine("API Error: Your request is not successful. If you are not able to debug this error properly, mail us at support@freshdesk.com with the follwing X-Request-Id");
                    Console.WriteLine("X-Request-Id: {0}", ex.Response.Headers["X-Request-Id"]);
                    Console.WriteLine("Error Status Code : {1} {0}", ((HttpWebResponse)ex.Response).StatusCode, (int)((HttpWebResponse)ex.Response).StatusCode);
                    using (var stream = ex.Response.GetResponseStream())
                    using (var reader = new StreamReader(stream))
                    {   
                      Console.Write("Error Response: ");
                      Console.WriteLine(reader.ReadToEnd());
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine("ERROR");
                    Console.WriteLine(ex.Message);
                }
            }

        }

    }
}
